src/cow-react/modules/wallet/web3-react/connectors/Injected/index.tsx

```diff
@@ -0,0 +1,204 @@
+  actions: Actions
+  onError?: () => void
+  walletUrl: string
+  searchKeywords: string[]
+}
+
+type Window = typeof Window & {
+  ethereum?: InjectedWalletProvider
+  injectedwallet?: InjectedWalletProvider
+}
+
+export class InjectedWallet extends Connector {
+  provider?: InjectedWalletProvider
+  walletUrl: string
+  searchKeywords: string[]
+
+  private get connected() {
+    return !!this.provider?.isConnected()
+  }
+
+  constructor({ actions, onError, walletUrl, searchKeywords }: injectedWalletConstructorArgs) {
+    super(actions, onError)
+
+    this.walletUrl = walletUrl
+    this.searchKeywords = searchKeywords
+
+    this.detectProvider()
+  }
+
+  public activate(desiredChainIdOrChainParameters?: number | AddEthereumChainParameter): Promise<void> | undefined {
+    if (!this.provider) {
+      window.open(this.walletUrl, '_blank')
+      return
+    }
+
+    const desiredChainId =
+      typeof desiredChainIdOrChainParameters === 'number'
+        ? desiredChainIdOrChainParameters
+        : desiredChainIdOrChainParameters?.chainId
+
+    if (this.connected && desiredChainId && desiredChainId === this.parseChainId(this.provider.chainId)) {
+      const desiredChainIdHex = `0x${desiredChainId.toString(16)}`
+
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      return this.provider
+        .request<void>({
+          method: 'wallet_switchEthereumChain',
+          params: [{ chainId: desiredChainIdHex }],
+        })
+        .catch(async (error: ProviderRpcError) => {
+          console.log('err!', error)
+
+          if (error.code === 4902 && typeof desiredChainIdOrChainParameters !== 'number') {
+            // if we're here, we can try to add a new network
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            return this.provider!.request<void>({
+              method: 'wallet_addEthereumChain',
+              params: [{ ...desiredChainIdOrChainParameters, chainId: desiredChainIdHex }],
+            })
+          }
+
+          throw error
+        })
+    }
+
+    return Promise.all([
+      this.provider.request({ method: 'eth_chainId' }) as Promise<string>,
+      this.provider.request({ method: 'eth_requestAccounts' }) as Promise<string[]>,
+    ]).then(([chainId, accounts]) => {
+      const receivedChainId = this.parseChainId(chainId)
+
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      if (!desiredChainId || desiredChainId === receivedChainId) {
+        return this.actions.update({ chainId: receivedChainId, accounts })
+      }
+
+      const desiredChainIdHex = `0x${desiredChainId.toString(16)}`
+
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      return this.provider!.request<void>({
+        method: 'wallet_switchEthereumChain',
+        params: [{ chainId: desiredChainIdHex }],
+      }).catch(async (error: ProviderRpcError) => {
+        console.log('err!', error)
```

Comment:
Double logging of error.
Please, add more understandable error text, like `[Injected provider] Chain siwtching error`

Reply:
Removed this